state/state.go

```diff
@@ -908,12 +907,14 @@ type CharmRefFull interface {
 	Actions() *charm.Actions
 	Config() *charm.Config
 	Revision() int
+	URL() string
```

Comment:
This move allows us to call `AddApplication()` with a charm that isn't retrieved from state but instead is created from the archive, as we have during bootstrap.